File: app/src/main/java/com/konektedi/vs/news/NewsViewModel.java
package com.konektedi.vs.news;


import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.konektedi.vs.utilities.NetworkStatus;

import java.util.List;

public class NewsViewModel extends AndroidViewModel {

    private LiveData<List<NewsModel>> news;
    private LiveData<NetworkStatus> networkStatus;


    public NewsViewModel(@NonNull Application application) {
        super(application);
        NewsRepository repository = new NewsRepository();
        news = repository.getNews();
        networkStatus = repository.networkState;
    }

    LiveData<List<NewsModel>> getAllNews() {
        return news;
    }

    LiveData<NetworkStatus> getNetworkStatus() {
        return networkStatus;
    }
}

File: app/src/main/java/com/konektedi/vs/news/NewsViewModel.kt
package com.konektedi.vs.news

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.konektedi.vs.utilities.common.Constants
import com.konektedi.vs.utilities.common.NetworkState
import com.konektedi.vs.utilities.models.Post
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class NewsViewModel : ViewModel() {
    var networkState: LiveData<NetworkState>
    private val executor: Executor
    private val tDataSource: LiveData<NewsDataSource>
    private val dataFactory: NewsDataFactory

    init {
        executor = Executors.newFixedThreadPool(5)
        dataFactory = NewsDataFactory(executor)

        tDataSource = dataFactory.mutableLiveData

        networkState = Transformations.switchMap(dataFactory.mutableLiveData) {
            it.networkState
        }
    }

    val news: LiveData<PagedList<Post>>
        get() {
            val pagedListConfig = PagedList.Config.Builder()
                    .setEnablePlaceholders(true)
                    .setInitialLoadSizeHint(Constants.INITIAL_LOAD)
                    .setPageSize(Constants.PAGE_SIZE)
                    .build()

            return LivePagedListBuilder(dataFactory, pagedListConfig)
                    .setFetchExecutor(executor)
                    .build()
        }
}